### DISCUSSION TITLE
Longitude coordinates appear erroneously

### DISCUSSION THREAD

**Discussion by itspangler** [allmaps/allmaps#92] [Ideas for Data Model, API and Backend]
A student came across this error while I was guest lecturing in a class last week. The `288` coordinates should all be somewhere around `-70`. I've no idea how the student even generated coordinates this way, and I've never seen it before, but I thought it should be on your radar.

Someone deleted the pixel mask containing the erroneous coordinates, but this is the georeference annotation that it would be: https://annotations.allmaps.org/images/bfc9c45499d14dc5

<img width="1440" alt="Screen Shot 2023-04-12 at 3 17 08 PM (1)" src="https://user-images.githubusercontent.com/39773984/232863317-0a189ab6-08d0-43ce-8db5-b775abcd94ce.png">

**Comment by sammeltassen** [comment]
This happens after panning the map to the left or right and placing the point there (subtracting or adding 360 degrees). This behavior can perhaps be useful when georeferencing maps that cross the 180th meridian, because otherwise points might be rendered in the wrong place. I have not yet tested this. In any case, clients should not have any issue accepting the "wrong" coordinates, see https://macwright.com/2015/03/23/geojson-second-bite#the-180th-meridian

**Comment by sammeltassen** [comment]
Initial test of a map crossing the 180th meridian:

Source:
- [Map of the Ocean Showing the Different Routes of the Navigators around the World](https://www.loc.gov/item/2021668634/) (Library of Congres)

In Allmaps:
- [Editor](https://editor.allmaps.org/#/collection?url=https%3A%2F%2Fwww.loc.gov%2Fitem%2F2021668634%2Fmanifest.json)
- [Viewer](https://viewer.allmaps.org/?url=https%3A%2F%2Fannotations.allmaps.org%2Fmanifests%2F32702348a216aac6)

Geojson.io
- https://geojson.io/#data=data:text/x-url,https://annotations.allmaps.org/manifests/32702348a216aac6.geojson

## Screenshots
The longitude values have huge consequences for the rendering of the map. For example:

These "correct" coords (lng, lat):

```js
[146.53438, -43.51173],
[174.82756, -41.30667],
[-65.12569, -54.66179],
[-73.53062, -41.794]
```

<img width="1464" alt="Screenshot 2024-03-22 at 12 10 35" src="https://github.com/allmaps/allmaps/assets/747644/0332d84a-3790-4d61-9884-f3ed82d944de">
Will mirror the map:
<img width="1464" alt="Screenshot 2024-03-22 at 12 10 52" src="https://github.com/allmaps/allmaps/assets/747644/f874f2de-d048-46ab-84c4-445b214543ee">

While these "incorrect" coords (lng, lat):

```js
[146.53438, -43.51173]
[174.82756, -41.30667]
[286.45312, -41.79414] \\ Invalid lng
[294.83624, -54.65821] \\ Invalid lng
```

Will render the map in the right place:

<img width="1464" alt="Screenshot 2024-03-22 at 12 17 52" src="https://github.com/allmaps/allmaps/assets/747644/eed2c838-1303-4f20-956b-beb2add1b825">

In other words we need the "incorrect" coords to correctly render maps crossing the 180th meridian, because it indicates the right direction to the client. But users are not aware of this when using the Editor and might accidentally pick points "on another globe" (+360 or -360) while panning the map.

**Comment by mclaeysb** [comment]
Thanks for this detailed analysis, @sammeltassen! 

It is indeed not possible to store point coordinates with longitudes outside the -180 to 180 range. [The GeoJSON Spec has some details](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.9) about how to handle features that cross the anti-meridian, but this only applies to lines or polygons, but GCPs are points and (multi)points crossing it are not stored in a specific way.

I see some possible solutions:
- Inferring possible anti-meridian crossing by testing for it and seeing if this reduces the error on the map. This is not easy because you also have to guess _which_ points could lie accross the antimeridian (and then combinatorics come into play).
- Adding a tag to a point that was logged accross the anti-meridian (e.g. +1 or -1). This requires correct tracking of the anti-meridian in the editor. It also ultimately means modifying the Annotation specification. Internally we could then treat the point with 'invalid' coordinates' (and at every export point return it to a valid coordinate plus a tag if possible).

**Reply by sammeltassen** [reply]
Actually we can infer the anti-meridian crossing from the resource coordinates (unless the original resource is mirrored...). So we can always export valid coordinates from the Editor, and generate "invalid" coordinates when a crossing is detected during the parsing of an annotation.

For the converted geojson mask we could follow the geojson spec, although it could result in chopped chunks on both sides of the map.
